### ISSUE TITLE
Restricted ImageNet Class Imbalance

### ISSUE THREAD

**Issue by inkawhich** [MadryLab/robustness#3]
Hello, I was wondering if you consider the class imbalance problem that is created in the Restricted ImageNet dataset when training the models? For example, the "Dog" superclass is made of 117 subclasses whereas the "Frog" class is made of only 3 subclasses. I don't see any special samplers or weighting in the loss functions but I may have missed something. Or, is this not a problem at all and you found that training as normal even with the class imbalance works fine?

Any response is appreciated,
Nate

**Comment by TLMichael** [comment]
I think the authors use `Restricted ImageNet` just for qualitative analysis, not quantitative analysis. So the class imbalance problem is not significant. But if we try to compare different models by clean or adversarial accuracy, the problem is not to be neglected.

Moreover, I found that the authors have another draft code(see [robustness_lib](https://github.com/MadryLab/robustness_lib)), where a dataset called `RestrictedImageNetBalanced` has been defined.

But another problem arises. `RestrictedImageNetBalanced` dataset has 14 classes, each class is made of 5 subclasses. So the total sample size is about 14*5000=70000. I think such a small sample size has the risk of overfitting. Are there any better configurations?

**Comment by inkawhich** [comment]
The balanced version that you linked to seems interesting. It wouldn't be that hard to extend it to 20 superclasses with 5 subclasses each so it uses 1/10th of the full dataset (e.g. other classes may be Cars, Musical_Instruments, Snakes, etc.). As for the overfitting, people are using resnet and wideresnet for cifar10 which is an all around smaller problem that the RestrictedImageNetBalanced. I guess I am not as concerned about the overfitting part but more about the biasing part.

**Comment by TLMichael** [comment]
That looks good! I could try both balanced datasets to figure out which one is right for me.

Does the bias problem arise from too many animal classes?
Maybe for a neural network, they are too different to induce a concept of animal. The difference between dog and fish may be as large as it between dog and car.

**Comment by andrewilyas** [comment]
Hi all! You're right that Restricted ImageNet has a big class imbalance, but as pointed out earlier, we are mostly using this dataset for qualitative analysis rather than quantitative, so it shouldn't be a problem.

The main bottleneck for the balanced dataset is just not enough data---using full sets instead of continuous ranges is an interesting direction for alleviating this though!

**Comment by TLMichael** [comment]
Full ImageNet adversarial training is still challenging, 
even with the advent of some techniques like [1] and [2].

They seem to have worse generalization performance than standard adversarial training.

[1] Shafahi, Ali, et al. "Adversarial Training for Free!." arXiv preprint arXiv:1904.12843 (2019).
[2] Zhang, Dinghuai, et al. "You Only Propagate Once: Accelerating Adversarial Training via Maximal Principle." arXiv preprint arXiv:1905.00877 (2019).

**Comment by andrewilyas** [comment]
Yeah, getting fully robust models on ImageNet is pretty hard (which is the main reason we use a subset of ImageNet instead of the whole thing)---on RestrictedImageNet though we can get very high robust and standard accuracy with adversarial training.

**Comment by TLMichael** [comment]
Thanks for your explanation. I've been wondering how much time you spent training a robust(PGD-7) ResNet-50 on RestrictedImageNet and how many GPUs you used.

Your information will be of great help to me. It helps my budget of device.

**Comment by andrewilyas** [comment]
If I'm remembering correctly, it usually takes a few hours on 2-3 1080Ti GPUs.

**Comment by TLMichael** [comment]
Thanks a lot!

**Comment by iamgroot42** [comment]
@andrewilyas In the "Adversarial Examples Are Not Bugs, They Are Features", it seems that metrics have indeed been reported and discussed for models trained on the Restricted ImageNet dataset. I realize that the paper's focus is not on getting state-of-the-art performance on this dataset but rather study the high-level patterns and phenomenon. However, I cannot stop thinking about how fixing the class imbalance (via weighted sampling or under-sampling, or any other method) would change the results?

**Comment by andrewilyas** [comment]
That's true---however its important to note that (a) the accuracy we get on Restricted-ImageNet is far, far better than what you would get with random chance for the D_R and D_NR dataset, and (b) for the D_det dataset, note that the "baseline" accuracy (i.e. what we would intuitively expect) is actually *0%*, as the images are consistently mislabeled---class imbalance should actually *hurt* you here (in the sense that during training you see 50% cats where the dataset originally has 50% dogs). Does this make sense?

We also have results on CIFAR-10 that further confirm the hypothesis.

**Comment by iamgroot42** [comment]
Yes: my point was that performance might (and probably would) **increase** if the imbalance were fixed, further reinforcing the claims based on empirical results :)

P.S. On a side note @andrewilyas , is the code to reproduce those experiments in the paper available anywhere? I could only find a link to the final datasets in the paper.

**Comment by andrewilyas** [comment]
There's no packaged code release---in making the results for the paper, we just used the adversarial attack functionality from this library, and also the training functionality from this library for D_det and D_rand. For D_R and D_NR, they are created through representation inversion, which we have released code for here: https://github.com/MadryLab/robust_representations/blob/master/image_inversion.ipynb

**Comment by inkawhich** [comment]
Ok thanks guys, so I guess the moral of the story is if you want quantitative results it may be useful to deal with the class imbalance problem but if you want mostly qualitative results the given RestrictedImageNet will work fine.
